File: frontend/src/api/channels.js
const BASE = "/api";

export async function fetchChannels({ query, category, country, language, liveOnly, status, page = 1, perPage = 40 }) {
  const params = new URLSearchParams();
  if (query) params.set("query", query);
  if (category) params.set("category", category);
  if (country) params.set("country", country);
  if (language) params.set("language", language);
  if (liveOnly) params.set("live_only", "true");
  if (status) params.set("status", status);
  params.set("page", page);
  params.set("per_page", perPage);

  const res = await fetch(`${BASE}/channels?${params}`);
  if (!res.ok) throw new Error("Failed to fetch channels");
  return res.json();
}

export async function fetchCategories() {
  const res = await fetch(`${BASE}/categories`);
  if (!res.ok) throw new Error("Failed to fetch categories");
  return res.json();
}

export async function fetchCountries() {
  const res = await fetch(`${BASE}/countries`);
  if (!res.ok) throw new Error("Failed to fetch countries");
  return res.json();
}

export async function fetchStats() {
  const res = await fetch(`${BASE}/stats`);
  if (!res.ok) throw new Error("Failed to fetch stats");
  return res.json();
}

export async function fetchStreams(channelId) {
  const res = await fetch(`${BASE}/channels/${channelId}/streams`);
  if (!res.ok) throw new Error("Failed to fetch streams");
  return res.json();
}

export async function runHealthCheck(channelId) {
  const res = await fetch(`${BASE}/healthcheck/${channelId}`, { method: "POST" });
  if (!res.ok) throw new Error("Health check failed");
  return res.json();
}

File: backend/app/routers/channels.py
import math

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.schemas import (
    ChannelOut, ChannelSearchParams, PaginatedChannels, StreamOut
)
from app.services.channel_service import (
    search_channels, get_channel_by_id, get_channel_streams
)

router = APIRouter(prefix="/api/channels", tags=["channels"])


@router.get("", response_model=PaginatedChannels)
async def list_channels(
    query: str | None = Query(None, description="Search term"),
    category: str | None = Query(None),
    country: str | None = Query(None),
    language: str | None = Query(None),
    live_only: bool = Query(False, description="Only show channels with a stream"),
    status: str | None = Query(None, description="Filter by health status: verified, live, or hide_offline"),
    page: int = Query(1, ge=1),
    per_page: int = Query(40, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
):
    params = ChannelSearchParams(
        query=query, category=category, country=country,
        language=language, live_only=live_only, status=status,
        page=page, per_page=per_page,
    )
    channels, total = await search_channels(db, params)
    return PaginatedChannels(
        channels=[ChannelOut.model_validate(c) for c in channels],
        total=total,
        page=page,
        per_page=per_page,
        total_pages=math.ceil(total / per_page) if total else 0,
    )


@router.get("/{channel_id}", response_model=ChannelOut)
async def get_channel(channel_id: str, db: AsyncSession = Depends(get_db)):
    channel = await get_channel_by_id(db, channel_id)
    if not channel:
        raise HTTPException(status_code=404, detail="Channel not found")
    return ChannelOut.model_validate(channel)


@router.get("/{channel_id}/streams", response_model=list[StreamOut])
async def get_streams(channel_id: str, db: AsyncSession = Depends(get_db)):
    streams = await get_channel_streams(db, channel_id)
    return [StreamOut.model_validate(s) for s in streams]
